[IMP] helpdesk_mgmt_sla: implement more advanced rules

BojanOD opened the pull request:
Implement more advanced rules to filter the tickets it applies to.
Additionally, the stage_id field is removed as it wasn't used.

pedrobaeza commented:
Why a different module instead of integrating this into SLA? What are the improvements (the README is very poor)?

damdam-s commented on helpdesk_mgmt_sla_rule/models/helpdesk_ticket.py:
```diff
@@ -0,0 +1,83 @@
+class HelpDeskTicket(models.Model):
+    _inherit = "helpdesk.ticket"
+
+    sla_id = fields.Many2one("helpdesk.sla", string="SLA", compute="_compute_sla")
```
missing store=True to be able to search on this field (like you do here https://github.com/OCA/helpdesk/pull/362/files?diff=split&w=0#diff-5e554a183028bdbcbc1eb0a377cf2553784b63af8c77381f454c5ab1eb875c8eR11)

damdam-s commented on helpdesk_mgmt_sla_rule/models/helpdesk_sla.py:
```diff
@@ -0,0 +1,13 @@
+    _inherit = "helpdesk.sla"
+
+    def check_sla(self):
+        tickets = self.env["helpdesk.ticket"].search([("sla_id", "!=", False)])
```
1. add comment to explain why this is a full override of the method
2. you can't currently search on this field because the `sla_id` field is not stored

damdam-s commented on helpdesk_mgmt_sla_rule/models/helpdesk_ticket.py:
```diff
@@ -0,0 +1,83 @@
+    sla_id = fields.Many2one("helpdesk.sla", string="SLA", compute="_compute_sla")
+
+    def check_sla_rule_m2o(self, rule, check, field):
+        if not rule[field] or not check:
+            return True
+        elif self[field].id == rule[field].id:
+            return True
+        else:
+            return False
```
```suggestion
        return (not rule[field] or not check) or (self[field].id == rule[field].id)
```
this should work to replace your tests

damdam-s commented on helpdesk_mgmt_sla_rule/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1 @@
+This module improves the check sla by adding a sla.rule model.
```
add a better description mentionning this module add a kind of rule engine for helpdesk SLA

add a usage file to explain how to use it and configure it

damdam-s commented on helpdesk_mgmt_sla_rule/security/ir.model.access.csv:
```diff
@@ -0,0 +1,2 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_sla_rule,access.sla.rule,model_sla_rule,helpdesk_mgmt.group_helpdesk_manager,1,1,1,1
```
missing rule for 'simple' user to be able to read the rules

BojanOD commented:
> Why a different module instead of integrating this into SLA? What are the improvements (the README is very poor)?

Hi @pedrobaeza , I wrote down some text in the README and DESCRIPTION file.

pedrobaeza commented:
Thanks for the README update, although that one is supposed to be user oriented documentation, not technical design one. I still think the rules can be part of the main module, and only develop and extension module for including the things related with the types.

elvise commented:
@BojanOD @pedrobaeza why not put there the domain (widget) ?

BojanOD committed:
[IMP] helpdesk_mgmt_sla: implement more advanced rules
Implement more advanced rules to filter the tickets it applies to.
Additionally, the stage_id field is removed as it wasn't used.

BojanOD commented:
Hi @pedrobaeza @elvise , I made and update to the PR.
It would be nice if you can have a look.
Thanks

pedrobaeza commented on helpdesk_mgmt_sla/models/helpdesk_sla.py:
```diff
@@ -12,20 +13,49 @@ class HelpdeskSla(models.Model):
+        comodel_name="helpdesk.ticket.category", string="Categories"
+    )
+    tag_ids = fields.Many2many(comodel_name="helpdesk.ticket.tag", string="Tags")
+    stage_ids = fields.Many2many(comodel_name="helpdesk.ticket.stage", string="Stages")
```
That this need a migration script stage_id to stage_ids?

ivantodorovich replied:
AFAICS the `stage_id` field wasn't really used in the code before 🤔 I'm not sure what purpose it had

ivantodorovich commented on helpdesk_mgmt_sla/views/helpdesk_sla_views.xml:
```diff
@@ -14,17 +14,24 @@
                             />
                             <separator colspan="2" string="Target" class="mt32" />
-                            <field name="stage_id" string="Reach Stage" />
+                            <field name="stage_ids" string="Reach Stages" />
```
```suggestion
                            <field name="stage_ids"  />
```
And move it before `<field name="domain" />`, because it's part of the filters

elvise commented:
> Details

As the user, i would like use the domain based on ticket form:
![Schermata 2022-09-08 alle 09 36 06](https://user-images.githubusercontent.com/5269577/189063787-e233ed35-9a3a-41df-912d-4f2acd28fca6.png)

This because i guess it's more useful, what do you think ?

elvise commented:
@BojanOD i would like improve just a little bit the UX:
![Schermata 2022-09-08 alle 09 41 03](https://user-images.githubusercontent.com/5269577/189065571-1cdd872a-67db-457f-ae65-056c728848d2.png)

Looks reasonable ?

BojanOD commented:
Hi @ivantodorovich @pedrobaeza @elvise ,

I made the necessary changes. Can you look into the code, and give me your thoughts if I need to change/add anything.

pedrobaeza commented on helpdesk_mgmt_sla/models/helpdesk_sla.py:
```diff
@@ -3,29 +3,61 @@
     days = fields.Integer(string="Days", default=0, required=True)
     hours = fields.Integer(string="Hours", default=0, required=True)
-    note = fields.Char(string="Note")
+    note = fields.Html(string="Note")
```
This need a text2html conversion in a migration script.

pedrobaeza reviewed (approved):
Very good addition, thanks!

Code review.

elvise commented:
> Hi @ivantodorovich @pedrobaeza @elvise ,
> 
> 
> 
> I made the necessary changes. Can you look into the code, and give me your thoughts if I need to change/add anything.

@BojanOD thanks so much! 

If is not to much these are my last 2 cents :) 

For "Archive" please follow the standard UI

If remove the list stages then we can move the field in the left side 
![image](https://user-images.githubusercontent.com/5269577/189197725-6873974d-e6f3-44df-95b3-e3d79d34903a.jpeg)

Add "Schedule Activity" 
![image](https://user-images.githubusercontent.com/5269577/189198595-a99ebf2f-f244-405b-9228-2c0b1e6f7f5c.jpeg)

Summernote full form
![image](https://user-images.githubusercontent.com/5269577/189198615-9d67f376-bf79-43d5-888e-cd8bb4b57778.jpeg)

BojanOD commented:
@elvise I made the requested changes. Can you review my code. Thanks

elvise commented:
> @elvise I made the requested changes. Can you review my code. Thanks

@BojanOD Thanks for your changes! 
I guess for the final user it's more clear with "Archive" 
What do you think ? 

![image](https://user-images.githubusercontent.com/5269577/189897285-eeaf52f1-4a40-49e7-852b-f2d9ec7ea5f4.jpeg)

pedrobaeza commented:
In this version smart buttons for this area removed by convention and leave Action > Archive

elvise commented:
> In this version smart buttons for this area removed by convention and leave Action > Archive

As you want, for me both of them are good, my concern is that there is always good UI and UX ✌🏻

BojanOD commented:
@elvise I made the necessary change, you can have a look.

elvise commented:
> @elvise I made the necessary change, you can have a look.

@BojanOD Wooooow !

P E R F E C T 💣

elvise reviewed (approved):
Great Job 👏

pedrobaeza commented:
/ocabot merge major
